Add package linux-firmware and use it on metal-* variants

foersleo opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #3246

**Description of changes:**

With introduction of Metal variants we opened up Bottlerocket to a wider variety of Hardware configurations. In the past we have added driver support for some server grade hardware, mostly in the 10G+ NIC realm. While adding the drivers, we overlooked that some of these devices may need firmware binaries to be available in order to update firmware and function with the driver version shipped with the kernel.

Add firmware from [linux-firmware](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/) to our metal variants for drivers we support in Bottlerocket. I have made an effort to cut down the list of binaries to only the ones we currently supply drivers for, as the linux-firmware package is quite large in full. I was able to cut the size of the installed size on disk down from 426 MB for the full package to 29 MB for the stripped down version carrying only the needed firmware.

We could strip the size on disk down even further when using compressed firmware images. Currently ZSTD and XZ compression are supported. While ZSTD does not change the size significantly XZ compression can further reduce the size from 29 MB (uncompressed) to 25 MB (xz compressed). However, doing that requires us to enable additional kernel config options `CONFIG_FW_LOADER_COMPRESS` and `CONFIG_FW_LOADER_COMPRESS_XZ`. That optimization can probably be part of a separate **commit.**

The resulting firmware package carries firmware for the following list of drivers and their corresponding config options we have configured for our metal kernels:

```
cxgb4 - CONFIG_CHELSIO_T4
tg3 - CONFIG_TIGON3
bnx2x - CONFIG_BNX2X
netxen_nic - CONFIG_NETXEN_NIC
qed - CONFIG_QED
myri10ge - CONFIG_MYRI10GE
i915 - CONFIG_DRM_I915
ice - CONFIG_ICE
```

**Testing done:**

I have done some build testing. Unfortunately assessing if all the firmware is in the correct place and can be reached by the drivers is beyond what I can do as I do not have access to these devices.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
linux-firmware: Add base firmware package

Create linux-firmware package and import latest source from upstream.
Adjust spec file to package all firmware files available.

Signed-off-by: Leonard Foerster <foersleo@amazon.com>

Commit:
linux-firmware: Remove a lot of firmware we do not need in Bottlerocket

Bottlerocket is an OS with a comparably cut-down kernel config and as
such does not provide the wide range of drivers one can find in general
purpose linux distributions. Many of these devices, Bottlerocket does
not support have firmware binaries shipped through linux-firmware.

Drop those firmware blobs we do not ship drivers for in Bottlerocket. I
have tried to section the removal into groups of drivers with
commonalities (e.g. Wifi and Bluetooth devices). Each patch has a list
mapping each driver to a specific kernel config option so we can easily
find which firmware binaries to add back into Bottlerocket if we add
drivers.

The patches only touch the WHENCE file and not the actual binaries as
rpm does not like to apply git binary patches. The WHENCE file is a
record of all the firmware binaries in the package with additional
information for each file. On installation the copy-firmware.sh script
uses the information in the file to determine which files to copy, so it
is enough to remove information from WHENCE in order to remove binaries
from the installed directories.

Signed-off-by: Leonard Foerster <foersleo@amazon.com>

Commit:
variants: metal: Add linux-firmware package

With the metal-* variants we are targetting a wider variety of hardware
platforms. In order to enable users to properly use Bottlerocket on
platforms we provide drivers for we also need to supply firmware
binaries for some of these devices. Ship these in the metal variants
through new package linux-firmware.

Signed-off-by: Leonard Foerster <foersleo@amazon.com>

foersleo commented:
I have not put the source tar in the lookaside cache yet. Hence, the failing test runs. I want to check the licensing on these firmware blobs we ship first. The linux-firmware package is an interesting place with regards to licenses.

Edit: So all the binaries that will currently go into this package are marked as Redistributable with no modification permitted.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,89 @@
+Version:    20230625
+Release:    1%{?dist}
+Summary:    Firmware files used by the Linux kernel
+License:    GPL+ and GPLv2+ and MIT and Redistributable, no modification permitted
```
This needs to be a valid SPDX expression with SPDX license identifiers.

foersleo replied:
So if I am reading the SPDX grammar and License list correctly this would be something like this:

```
GPL-1.0-or-later AND GPL-2.0-or-later AND MIT AND ?
```

A lot of the firmware is licensed as "Redistributable, no modification permitted" (or that is at least the name for that bucket that fedora uses. All the firmware we currently have left to include falls into that group. If I understand correctly there is currently no solution for these in the SPDX guidelines. The [Fedora licensing guide](https://docs.fedoraproject.org/en-US/legal/update-existing-packages/#_redistributable_no_modification_permitted) marks usage of that with SPDX as unknown.

What do can we make of this?

I can see that for some of the binaries we include (if we take this PR) we might be able to do something like `LicenseRef-scancode-chelsio-linux-firmware` (according to https://scancode-licensedb.aboutcode.org/). I will see if I can find that for all of them.

foersleo replied:
So I looked at the `LICEN[CS]E` files we ship with what is proposed here. That is the following list with a mapping of what I think is the right License to specify if I have found a fitting specifier or leaving a comment on the state.

```
GPL-2 - clear what to do
GPL-3 - clear what to do
LICENCE.chelsio_firmware - LicenseRef-scancode-chelsio-linux-firmware 
LICENCE.myri10ge_firmware - seems to be no available
LICENCE.phanfw - LicenseRef-scancode-qlogic-firmware is identical
LICENSE.i915 - seems to be not available and not matching any other intel license
LICENSE.ice - seems to be not available and not matching any other intel license
LICENSE.ice_enhanced - seems to be not available and not matching any other intel license
```

Refering to the lists of Licenses in https://spdx.org/licenses/ and https://scancode-licensedb.aboutcode.org/

foersleo replied:
I have spend some more time on this and found some scancode licenses that seem to match the ones for Intel (i915, ice, ice_enhanced). For the Myricom license there is have still not found a proper License identifier to put here without hesitancy.

Talked to Ben about this and will reach out to licensing experts in our OSPO to help resolve these questions and be on the safe side.

foersleo replied:
After reaching out to ospo and playing around some more with the scancode tooling I have come up with a list of SPDX license reference codes. I think we are fine with those, although I am not super happy about `LicenseRef-scancode-free-unknown` for the license of `tg3` as specified in `WHENCE`.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,89 @@
+License:    GPL+ and GPLv2+ and MIT and Redistributable, no modification permitted
+URL:        https://www.kernel.org/
+
+Source0:    https://www.kernel.org/pub/linux/kernel/firmware/linux-firmware-%{version}.tar.xz
```
Are there any dependencies between these releases and the particular kernel versions?

foersleo replied:
That is a good question. There should not be a dependency of a certain firmware package state requiring a certain kernel version. The worst that could happen there would be additional firmware files the kernel does not know about. Or installing compressed firmware the kernel does not know how to decompress (See comment in spec file).

The other way around we may get into situations where a driver expects certain firmware files which are not there in which case we need to update linux-firmware.

If we want/need a hard dependency between firmware packages is debatable I guess. Maybe make it a rule to just update it in lockstep with major kernel updates is an option?

markusboehme replied:
It would be drivers checking for firmware versions and features, so using newer firmware with older drivers would work. In other cases, drivers would explicitly request a particular firmware version to be loaded (and older firmware files be kept around in new releases). Leo's suggestion of updating with the introduction of a new kernel series at the latest seems sound to me, but more frequently would be fine, too.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,89 @@
+%global debug_package %{nil}
+
+%global _firmwarepath   /usr/lib/firmware
```
This should be under the sysroot path, otherwise it will conflict with our `/usr` symlink at build time.

```suggestion
%global fwdir %{_cross_libdir}/firmware
```

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,89 @@
+Name:       %{_cross_os}linux-firmware
+Version:    20230625
+Release:    1%{?dist}
+Summary:    Firmware files used by the Linux kernel
```
nit: I try to avoid the column-aligned layouts since those can lead to whitespace-related churn in future diffs, where other lines have to be adjusted to preserve the column

```suggestion
Name: %{_cross_os}linux-firmware
Version: 20230625
Release: 1%{?dist}
Summary: Firmware files used by the Linux kernel
```

foersleo replied:
I was wondering about that when looking at other specs, but muscle memory got the better of me. Fixed.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,89 @@
+Requires:   linux-firmware-whence
+Provides:   kernel-firmware = %{version}
+Obsoletes:  kernel-firmware < %{version}
+
```
We shouldn't need any of these - just put the "whence" related files in the same package as the others.
```suggestion
```

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,89 @@
+
+rm -f check_whence.py configure Makefile README
+
+#Remove executable bits from random firmware
```
```suggestion
# Remove executable bits from random firmware
```

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,89 @@
+sed -i -e 's:^./::' linux-firmware.{files,dirs}
+sed -i -e 's!^!/usr/lib/firmware/!' linux-firmware.{files,dirs}
+sed -i -e 's/^/"/;s/$/"/' linux-firmware.files
+sed -e 's/^/%%dir /' linux-firmware.dirs >> linux-firmware.files
```
I would just bulk-include everything in the firmware directory, possibly splitting it up by device or directory if it's not an overwhelming number. (Like the headers subpackage in the kernel packages, with `asm`, `xen` includes.)

foersleo replied:
I went for a blanket bulk include as suggested. Unfortunately the directory structure of linux-firmware is only existent for some vendors/firmware. For others it is a bit of a wild west setting. Going the bulk-include rule here puts more pressure on being clean in the patches, while keeping the rpm brief.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,89 @@
+# Cleanup unnecessary files
+pushd %{buildroot}/%{_firmwarepath}
+
+rm -f check_whence.py configure Makefile README
```
We could potentially just exclude these four files in the `%files` section.

foersleo replied:
These files are actually not there I found. I had oriented myself on Fedora/AL2023 spec file for linux-firmware. Removed.

foersleo commented:
:arrow_up: Force push was just rebasing on top of develop before tending comments.

foersleo commented:
:arrow_up: Force push tended most comments by Ben. The only one I have not fixed yet is the one about the SPDX license specs because of the above mentioned reasons.

I went also ahead and removed unneeded license files in the patches in commit 2 of the series. That helps uncluttering the license directory in the installation and only ships the license files for firmware we actually ship.

zmrow reviewed (approved):
From the variant perspective this looks good - I don't have any additional comments about the package or patches!

yeazelm commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,62 @@
+%license LICENCE.* LICENSE.* GPL* WHENCE
+%{_cross_attribution_file}
+
+%changelog
```
We recently noticed we don't need to include this empty `%changelog` in our specfiles. So you can drop it if you'd like.

foersleo replied:
Done.

yeazelm reviewed (commented):
I'm happy with the outcome of trimming this size down. I am curious if we know how difficult it will be to keep up with these updates as things are added to the `linux-firmware` bundle upstream. Is there anything special we need to track to ensure updates don't accidentally start bloating with more things or will we just need to be diligent to check each update doesn't include things we don't need at the time of reviewing the update?

foersleo commented:
> I'm happy with the outcome of trimming this size down. I am curious if we know how difficult it will be to keep up with these updates as things are added to the `linux-firmware` bundle upstream. Is there anything special we need to track to ensure updates don't accidentally start bloating with more things or will we just need to be diligent to check each update doesn't include things we don't need at the time of reviewing the update?

That is a great question and I am a bit concerned about that myself. In all honesty I do not know yet. It looks like there is a release for linux-firmware about every 4-6 weeks (roughly once a month). I will have a stab at what it takes to update once the next release is out.

In the extreme case where these updates are too burdensome to do every month, we could just go with a linux-firmware update when we introduce a new major kernel. As discussed in a comment to Ben's question about dependency to specific kernel versions, the firmware images would be fine for the kernels we currently support. I do not expect major driver updates that would rely on newer firmware files (The file names are usually hard-coded in the drivers) within a stable series.

In an ideal world we would spend some time to build tooling to automatically strip down the files according to some kernel config file. Given the structure of the firmware repo I'd assume this to be quite challenging with lots of corner cases. But it is something we might want to try our hands on when we find some time

foersleo commented:
Firmware was confirmed to be at the right spot and the driver finding it properly by Max Böhm https://github.com/bottlerocket-os/bottlerocket/issues/3246#issuecomment-1656523329

Thanks a lot Max for helping out here!

markusboehme commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,51 @@
+%prep
+tar -xof %{SOURCE0}
+%setup -TDn linux-firmware-%{version}
+%autopatch -p1
```
Nit: The sequence is a bit peculiar (telling `%setup` not to extract Source0 only to extract it manually before). This could become `%setup -n linux-firmware-%{version}`, or to also squash the patching, just `%autosetup -n linux-firmware-%{version} -p1`.

foersleo replied:
It is, and I believe I had a good reason when I first started down this journey of packaging linux-firmware. Not it is just lingering around from then without any reason. I confirmed that it can be simplified to the proposed shorter versions and creates an identical package (apart from expected differences like timestamps and buildids). I will fix that with a next iteration.

markusboehme commented:
I wanted to review those patches as applied to linux-firmware and there's a bit of an interesting problem there: They do apply fine for me via `patch` and `git apply`, but then I lose their history and commentary. When trying to apply them via `git am`, at least `0003-linux-firmware-bt-wifi-Remove-firmware-for-Bluetooth.patch` fails. The first failure is for the first line of the deleted `LICENSE.nxp` which does seem to use CRLF line endings. The patch seems to have converted those to just LF, and hence doesn't match. 

I know Git can perform conversion of line endings, but it seems unexpected here. Can you check whether the original patches as generated by `git format-patch` were correct and this only happened when committing the patch files to the Bottlerocket repo? If so, I wonder whether disabling EOL conversion for `*.patch` files via `.gitattributes` (by treating those as binary files) would be desirable.

For the review, I just pass `--ignore-whitespace` to `git am`.

markusboehme reviewed (commented):
Thanks for the diligent work combing through all this! I hope it all plays well with updates and won't need to be redone. :smiling_face_with_tear: 

I'm happy to approve once we reach a conclusion about the SPDX identifiers and the line endings (not caring at this point could be one).

foersleo commented:
> I wanted to review those patches as applied to linux-firmware and there's a bit of an interesting problem there: They do apply fine for me via `patch` and `git apply`, but then I lose their history and commentary. When trying to apply them via `git am`, at least `0003-linux-firmware-bt-wifi-Remove-firmware-for-Bluetooth.patch` fails. The first failure is for the first line of the deleted `LICENSE.nxp` which does seem to use CRLF line endings. The patch seems to have converted those to just LF, and hence doesn't match.
> 
> I know Git can perform conversion of line endings, but it seems unexpected here. Can you check whether the original patches as generated by `git format-patch` were correct and this only happened when committing the patch files to the Bottlerocket repo? If so, I wonder whether disabling EOL conversion for `*.patch` files via `.gitattributes` (by treating those as binary files) would be desirable.
> 
> For the review, I just pass `--ignore-whitespace` to `git am`.

The line endings seems to be an interesting issue with `git format-patch` and `git am`. This is **not** an artifact of adding patches to a separate repository. The original patches do not apply to a clean checkout/unpack of linux firmware when trying to apply via `git am` without extra flags. We can work around this by passing flags `--ignore-whitespace` (as pointed out by you), or `--keep-cr` which will apply the patches, but create a warning about quoted CRLF:

```
$ for p in ../../linux-firmware-20230625/*.patch; do git am --keep-cr $p; done
Applying: linux-firmware: snd: remove firmware for snd/audio devices
Applying: linux-firmware: video: Remove firmware for video/broadcast devices
warning: quoted CRLF detected
Applying: linux-firmware: bt/wifi: Remove firmware for Bluetooth/WiFi devices
Applying: linux-firmware: scsi: Remove firmware for SCSI devices
Applying: linux-firmware: usb: remove firmware for USB/Serial/PCMCIA devices
Applying: linux-firmware: ethernet: Remove firmware for ethernet/IB devices
Applying: linux-firmware: Remove firmware for Accelarator devices
Applying: linux-firmware: gpu: Remove firmware for GPU devices
Applying: linux-firmware: various: Remove firmware for various devices
Applying: linux-firmware: amd-ucode: Remove amd microcode
```

[Stackoverflow](https://stackoverflow.com/questions/6289001/git-am-format-patch-control-format-of-line-endings) has some more detail about what is happening and what git tries to do to work around this. The quoted commit message in [one of the answers](https://stackoverflow.com/a/28257267) is particularly enlightening about the odds and ins of CRLF. The gist of which is: `git-am` by default does not trust line endings coming in through mail because SMTP transport is CRLF-unsafe.

Unfortunately the proposed `--transfer-encoding` option seems to be not available with `git format-patch` so it seems there is not really anything we can do about this on the patch creation side.

foersleo commented:
:arrow_up: Force push simplified `%prep` as suggested by @markusboehme and dropped `%changelog` as suggested by @yeazelm

markusboehme commented:
@foersleo Thanks for looking closer into it! TIL about `git am --keep-cr` which seems to be the better alternative.

foersleo commented:
Converting this one to draft for now to reduce risk of accidental merge while I am sorting out the questions regarding the Licenses.

Commit:
linux-firmware: Adjust Lincense definition

As we want to adhere to SPDX identifiers the Licensing originally given
by AL/Fedora is not suitable. Adjust the Licenses to cover the firmware
we actually package.

Signed-off-by: Leonard Foerster <foersleo@amazon.com>

foersleo commented:
:arrow_up: Force push included fixing up the list of SPDX identifiers as discussed with OSPO and added the `force-upstream` flag to the package as introduced in #3333 .
Unfortunately I did mix in some catch up with the upstream develop branch and so the comparison to the previous state is littered with a lot of other changes.

Will do an additional rebase shortly to base it on top of current  develop branch.

markusboehme reviewed (approved):
Still looking good to me with the updated SPDX license specifiers.

arnaldo2792 commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,67 @@
+Patch0010: 0010-linux-firmware-amd-ucode-Remove-amd-microcode.patch
+
+%description
+This package includes firmware files required for some devices to operate.
```
nit: for consistency with other packages, use `%{summary}.`

foersleo replied:
Fixed.

arnaldo2792 commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -0,0 +1,67 @@
+
+# Here we have potential to shave off some extra space by using `install-xz` of
+# `install-zst` to compress firmware images on disk. However, that functionality
+# relies on kernels being configured with `CONFIG_FW_LOADER_COMPRESS_[ZSTD|XZ]`
```
Should we have an issue to track enabling this config?

foersleo replied:
https://github.com/bottlerocket-os/bottlerocket/issues/3359

arnaldo2792 commented on packages/linux-firmware/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+package-name = "linux-firmware"
+
+[lib]
+path = "/dev/null"
```
We are not supposed to use this path, as explained in https://github.com/bottlerocket-os/bottlerocket/pull/3288

foersleo replied:
Thanks, I had not seen that one. Fixed.

foersleo commented:
:arrow_up: tended the comments mentioned by Arnaldo.

In addition I removed the force-upstream falg from the linux-firmware package. Discussing it with Ben we should be fine hosting a copy in the look-aside cache and we want to do that for as many packages as possible.

foersleo commented:
Uploaded the source artifacts to lookaside cache and checked the automated test builds succeed. Merging.
